vshard/storage/schema.lua

```diff
@@ -80,140 +94,51 @@ local function schema_init_0_1_15_0(username, password)
     bucket:create_index('pk', {parts = {'id'}})
     bucket:create_index('status', {parts = {'status'}, unique = false})
 
-    local storage_api = {
-        'vshard.storage.sync',
-        'vshard.storage.call',
-        'vshard.storage.bucket_force_create',
-        'vshard.storage.bucket_force_drop',
-        'vshard.storage.bucket_collect',
-        'vshard.storage.bucket_send',
-        'vshard.storage.bucket_recv',
-        'vshard.storage.bucket_stat',
-        'vshard.storage.buckets_count',
-        'vshard.storage.buckets_info',
-        'vshard.storage.buckets_discovery',
-        'vshard.storage.rebalancer_request_state',
-        'vshard.storage.rebalancer_apply_routes',
-    }
-    for _, name in ipairs(storage_api) do
-        box.schema.func.create(name, {setuid = true})
-        box.schema.user.grant(username, 'execute', 'function', name)
-    end
+    local exports = lvexports.log[1]
+    assert(exports.version == '0.1.15.0')
+    exports = lvexports.compile(exports)
+    lvexports.deploy_funcs(exports)
+    lvexports.deploy_privs(exports, username)
     box.space._schema:replace({'vshard_version', 0, 1, 15, 0})
```

Comment:
> Why don't we set _schema.vshard_version automatically?

What do you mean automatically? Where is it automatic?

> Currently if we manually updated to 0.1.16 and then we try to use automatic update, we'll try to deploy everything one more time.

The "users" (Alexander and Mergen - nobody else should use this, those exports are bad) should bump the version if they want to fully deploy the entire schema themselves. If they don't do that, then it will be done again inside vshard, yes (unless schema management mode is not `manual_access`). The patch had no intention to give the "users" a full featured upgrade module for deploying the whole schema. VShard still does things on its own, like `_bucket` space, for example. It can be probably extended in the future for deploying the entire schema, but that would be ridiculous IMO. Even the current requirement to make the exports manual is a sign of something being wrong in `config` module. I can understand the need to manage users on your own, but not the vshard private and public API. I will be against exporting the entire schema management as a public module if it ever comes to that.

> Maybe we can add exports.upgrade_callback() to the returned value of exports.compile? So, that user can manually call it, when upgrade is done.

Why would the "users" need a callback for that? Can they just do whatever they want after `deploy_funcs/privs()` returns?

Reply:
> VShard still does things on its own, like _bucket space, for example.

And that's pretty much it, nothing is done by vshard, when schema management mode is `manual`, except the space creation and writing `vshard_version` to `_schema`.

> Why would the "users" need a callback for that? Can they just do whatever they want after deploy_funcs/privs() returns?

I don't think we should export space creation, but `exports.upgrade_callback` could check, that the whole version was deployed (on 0.1.15 that space is also exists, on all versions - that functions were created) and write `vshard_version`.

> Even the current requirement to make the exports manual is a sign of something being wrong in config module.

Agree, I suppose, we should've stopped on the previous version, when only user wasn't created. IMO, granting privs should be done by vshard.

-----

In the end, I'm okay with the current version too

Reply:
> when schema management mode is manual

We only have `auto` and `manual_access`, no full `manual`.

> I don't think we should export space creation, but exports.upgrade_callback could check, that the whole version was deployed 

Nah, too complicated. The stuff they want to do is feasible with the current solution too. If they come with more specific requests, we can think of it.

> Agree, I suppose, we should've stopped on the previous version, when only user wasn't created.

Yes. But I don't want to argue anymore for now. There will be another fight if they will want more.